In [1]:
install.packages("devtools")

also installing the dependencies ‘pkgconfig’, ‘BH’, ‘sys’, ‘ini’, ‘tibble’, ‘lazyeval’, ‘later’, ‘askpass’, ‘highr’, ‘markdown’, ‘xfun’, ‘diffobj’, ‘clipr’, ‘curl’, ‘fs’, ‘gh’, ‘git2r’, ‘purrr’, ‘rematch2’, ‘rprojroot’, ‘whisker’, ‘yaml’, ‘processx’, ‘R6’, ‘rex’, ‘htmlwidgets’, ‘magrittr’, ‘crosstalk’, ‘promises’, ‘mime’, ‘openssl’, ‘prettyunits’, ‘xopen’, ‘brew’, ‘commonmark’, ‘knitr’, ‘Rcpp’, ‘stringi’, ‘stringr’, ‘xml2’, ‘brio’, ‘praise’, ‘ps’, ‘waldo’, ‘usethis’, ‘callr’, ‘covr’, ‘desc’, ‘DT’, ‘httr’, ‘memoise’, ‘pkgbuild’, ‘pkgload’, ‘rcmdcheck’, ‘remotes’, ‘roxygen2’, ‘rstudioapi’, ‘rversions’, ‘sessioninfo’, ‘testthat’, ‘withr’





The downloaded binary packages are in
	/var/folders/z4/zq1sx9z918l822zf41yhflvr0000gn/T//RtmpmNplLu/downloaded_packages


In [2]:
library("devtools")

Loading required package: usethis



In [4]:
install_github("https://github.com/xuan-chen/simpleDE")

✔  checking for file ‘/private/var/folders/z4/zq1sx9z918l822zf41yhflvr0000gn/T/RtmpmNplLu/remotes12d6e6038381/xuan-chen-simpleDE-42f6df9/DESCRIPTION’
─  preparing ‘simpleDE’:
✔  checking DESCRIPTION meta-information
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  building ‘simpleDE_0.1.0.tar.gz’
   


In [5]:
library(simpleDE)

In [7]:
install.packages("data.table", dependencies=TRUE)

also installing the dependencies ‘bit’, ‘R.oo’, ‘R.methodsS3’, ‘RcppCCTZ’, ‘RcppDate’, ‘tinytex’, ‘bit64’, ‘R.utils’, ‘xts’, ‘nanotime’, ‘zoo’, ‘rmarkdown’





  There is a binary version available but the source version is later:
          binary source needs_compilation
rmarkdown    2.3    2.5             FALSE


The downloaded binary packages are in
	/var/folders/z4/zq1sx9z918l822zf41yhflvr0000gn/T//RtmpmNplLu/downloaded_packages


installing the source package ‘rmarkdown’




In [8]:
library(data.table)

# "Differential Expression Analysis on Expression Data of Mouse Embryonic Stem Cells"

In [9]:
# load the dataset
dt = fread("/Users/sdeshpande/Desktop/gene_analysis/simpleDE/GSE5583.txt")
head(dt)

ID        WT1    WT2    WT3    KO1    KO2    KO3   
1 100001_at   11.5    5.6   69.1   15.7   36.0   42.0
2 100002_at   20.5   32.4   93.3   31.8   14.4   22.9
3 100003_at   72.4   89.0   79.2   80.5  130.1   86.7
4 100004_at  261.0  226.2  365.1  432.0  447.3  288.1
5 100005_at 1086.2 1555.6 1487.1 1062.2 1365.9 1436.2
6 100006_at   49.7   52.9   15.0   25.8   48.8   54.8

In [10]:
#Here we only have the expression data matrix. But we can still create a simpleDE list for further analysis. However, we can be add some sample information in sample_info as index for differential expression analysis.
# Create simpleDE
mtx = as.matrix(dt[,-1])
rownames(mtx) = dt$ID
colnames(mtx) = colnames(dt)[-1]
simple.mouse = Create_simpleDE(mtx)
str(simple.mouse)

List of 3
 $ data       :List of 1
  ..$ raw: num [1:12488, 1:6] 11.5 20.5 72.4 261 1086.2 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:12488] "100001_at" "100002_at" "100003_at" "100004_at" ...
  .. .. ..$ : chr [1:6] "WT1" "WT2" "WT3" "KO1" ...
 $ sample_info:'data.frame':	6 obs. of  1 variable:
  ..$ sample_id: chr [1:6] "WT1" "WT2" "WT3" "KO1" ...
 $ gene_info  :'data.frame':	12488 obs. of  1 variable:
  ..$ gene_id: chr [1:12488] "100001_at" "100002_at" "100003_at" "100004_at" ...


In [11]:
#Now we add genotype information into sample_info.

simple.mouse$sample_info["genotype"] = rep(c("WT", "KO"), each = 3)
str(simple.mouse)

List of 3
 $ data       :List of 1
  ..$ raw: num [1:12488, 1:6] 11.5 20.5 72.4 261 1086.2 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:12488] "100001_at" "100002_at" "100003_at" "100004_at" ...
  .. .. ..$ : chr [1:6] "WT1" "WT2" "WT3" "KO1" ...
 $ sample_info:'data.frame':	6 obs. of  2 variables:
  ..$ sample_id: chr [1:6] "WT1" "WT2" "WT3" "KO1" ...
  ..$ genotype : chr [1:6] "WT" "WT" "WT" "KO" ...
 $ gene_info  :'data.frame':	12488 obs. of  1 variable:
  ..$ gene_id: chr [1:12488] "100001_at" "100002_at" "100003_at" "100004_at" ...


In [12]:
#Then we perform filtering and normalization.
simple.mouse = Filter_Data(simple.mouse, verbose = F)
simple.mouse = Normalize_Data(simple.mouse, option = "size_factor")

In [13]:
#Now we can perform differential expression test on two genotypes.
de.geno = DEA_test(simple.mouse, group = "genotype", verbose = F)
head(de.geno)

gene      t          FC        log_FC      p_value   bonferroni
100001_at 100001_at  0.1148786 0.9199573 -0.08342801 0.9168154 1         
100002_at 100002_at -1.1125831 2.1157742  0.74942082 0.3726832 1         
100003_at 100003_at  1.1573839 0.8092836 -0.21160593 0.3500913 1         
100004_at 100004_at  1.5993997 0.7300839 -0.31459576 0.1875973 1         
100005_at 100005_at -0.4741657 1.0684729  0.06623047 0.6614367 1         
100006_at 100006_at  0.2620043 0.9088099 -0.09561935 0.8074091 1         
          fdr      
100001_at 0.9808352
100002_at 0.8026614
100003_at 0.7907215
100004_at 0.6904555
100005_at 0.9051315
100006_at 0.9551950

In [14]:
#Now we are only interesting in genes with absolute log_FC larger than 0.5 and p_value smaller than 0.5.
de.res = Get_DEG(de.geno, "log_FC", 0.5)
de.res = Get_DEG(de.res, "p_value", 0.05)
str(de.res)
head(de.res)

'data.frame':	510 obs. of  3 variables:
 $ log_FC : num  -0.883 -1.29 -0.663 0.725 0.898 ...
 $ p_value: num  7.06e-07 1.87e-06 6.91e-06 7.45e-06 1.19e-05 ...
 $ fdr    : num  0.00882 0.01169 0.02327 0.02327 0.02975 ...


log_FC     p_value      fdr        
100618_f_at -0.8825696 7.062020e-07 0.008819051
96920_at    -1.2898932 1.872452e-06 0.011691592
93140_at    -0.6631696 6.905019e-06 0.023272627
97181_f_at   0.7245365 7.454397e-06 0.023272627
96183_at     0.8982370 1.191285e-05 0.029753532
102332_at   -0.6022340 1.571138e-05 0.032700624

# "Comparison of Differnetial Expression Analysis on Simulated Data with DESeq2"

This tutorial will be analyzing a expression dataset (GSE5583). It contains an expression matrix of embryonic stem cells from 3 wild types and 3 HDAC knock out mice. Histone deacetylase 1 (HDAC1) is an enzyme asssocatied with gene activation and expression and it plays an important role at development and proliferation. Differential expression analysis will be performed here to seek for significant genes associated with HDAC knock out and hope to build up the underlying biological mechanism.

In [17]:
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


The downloaded binary packages are in
	/var/folders/z4/zq1sx9z918l822zf41yhflvr0000gn/T//RtmpmNplLu/downloaded_packages


In [18]:
BiocManager::install("DESeq2")

Bioconductor version 3.12 (BiocManager 1.30.10), R 4.0.3 (2020-10-10)

Installing package(s) 'BiocVersion', 'DESeq2'

also installing the dependencies ‘bitops’, ‘formatR’, ‘blob’, ‘plogr’, ‘colorspace’, ‘RCurl’, ‘GenomeInfoDbData’, ‘zlibbioc’, ‘matrixStats’, ‘lambda.r’, ‘futile.options’, ‘DBI’, ‘RSQLite’, ‘XML’, ‘xtable’, ‘farver’, ‘labeling’, ‘munsell’, ‘viridisLite’, ‘GenomeInfoDb’, ‘XVector’, ‘MatrixGenerics’, ‘DelayedArray’, ‘futile.logger’, ‘snow’, ‘AnnotationDbi’, ‘annotate’, ‘RColorBrewer’, ‘gtable’, ‘isoband’, ‘scales’, ‘S4Vectors’, ‘IRanges’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘BiocGenerics’, ‘Biobase’, ‘BiocParallel’, ‘genefilter’, ‘locfit’, ‘geneplotter’, ‘ggplot2’, ‘RcppArmadillo’





The downloaded binary packages are in
	/var/folders/z4/zq1sx9z918l822zf41yhflvr0000gn/T//RtmpmNplLu/downloaded_packages


installing the source package ‘GenomeInfoDbData’


Old packages: 'KernSmooth', 'codetools', 'nlme'



In [19]:
library(DESeq2)

Loading required package: S4Vectors

Loading required package: stats4

Loading required package: BiocGenerics

Loading required package: parallel


Attaching package: ‘BiocGenerics’


The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB


The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs


The following objects are masked from ‘package:base’:

    Filter, Find, Map, Position, Reduce, anyDuplicated, append,
    as.data.frame, basename, cbind, colnames, dirname, do.call,
    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,
    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,
    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,
    tapply, union, unique, unsplit, which.max, which.min



Attaching package: ‘S4Vectors’


The follow

Here we simulate a dataset containing 10000 genes expression data in 10 samples from two condition (5 treatment vs 5 control). Among those 10000 genes, 100 genes are differentially expressed genes, simulated from two negative binomial distribution with distinct means. 

In [20]:
set.seed(20191124)
mu = rpois(10000, 20)
size = 20
ctr_dt = matrix(NA, 10000, 5)
trt_dt = matrix(NA, 10000, 5)
for (i in 1:100){
  ctr_dt[i, ] = rnbinom(n = 5, size =  size, mu = mu[i])
  trt_dt[i, ] = rnbinom(n = 5, size =  size, mu = mu[i] + 40)
}
for (i in 101:10000){
  ctr_dt[i, ] = rnbinom(n = 5, size = size, mu = mu[i])
  trt_dt[i, ] = rnbinom(n = 5, size = size, mu = mu[i])
}
dt = cbind(ctr_dt, trt_dt)
colnames(dt) = c(sapply(1:5, function(x) paste0("ctr",x)), sapply(1:5, function(x) paste0("trt",x)))
rownames(dt) = c(sapply(1:100, function(x) paste0("DEG",x)), sapply(1:9900, function(x) paste0("non_DEG",x)))
dt[1:10,]
dt[101:110,]


ctr1 ctr2 ctr3 ctr4 ctr5 trt1 trt2 trt3 trt4 trt5
DEG1  28   22   31   28   25    82  99   70   68   58  
DEG2   8   18   10   19   12    56  66   51   49   34  
DEG3  33   20   42   16   25    82  70   62   99   64  
DEG4  26   24   14   24    8    44  77   70   58   56  
DEG5  20   25   13   25   29    57  76   66   87   41  
DEG6  25    8   23   36   19    48  67   69   55   74  
DEG7  13   23   24   16   23    59  69   56   49   63  
DEG8  13   14   17   19    7    59  80   53   54   60  
DEG9  13   24   15   22   21   101  68   83   48   71  
DEG10 25   24   23   42   25    49  75   54   50   65

ctr1 ctr2 ctr3 ctr4 ctr5 trt1 trt2 trt3 trt4 trt5
non_DEG1  10   22   18   23   20    7   20   17    8   15  
non_DEG2  11   12   18   19   24   18   12   16   17   10  
non_DEG3  42   25   28   25   21   26   31   35   18   33  
non_DEG4  29    9   26   25    9   16   16   17   23   15  
non_DEG5  12   17   16    8    7    6   16   15   18   12  
non_DEG6  26   33   34   54   36   35   40   29   45   63  
non_DEG7  15   29   24   23   20   23   24   18   14   17  
non_DEG8  26   32   27   27   17   22   28   26   16   28  
non_DEG9  16   13   13   18   19   21   12   12   13   17  
non_DEG10 34   23   42   22   23   16   23   30   28   22

In [21]:
# we also set up the sample information and gene information.
# Create simpleDE
mtx = dt
s_data = data.frame(treatment = rep(c(0,1), each = 5))
g_data = data.frame(is_DEG = c(rep(1, 100), rep(0, 9900)))

In [22]:
#Now we perform simpleDE and get the final differnetially expressed genes by threhold of 0.5 for log fold-change and 0.01 for p value.
simpleDE.test = function(){
  simple.dt = Create_simpleDE(mtx, s_data, g_data)
  simple.dt = Filter_Data(simple.dt, verbose = F)
  simple.dt = Normalize_Data(simple.dt, option = "size_factor")
  de.geno = DEA_test(simple.dt, group = "treatment", verbose = F)
  de.res = Get_DEG(de.geno, "log_FC", 0.5, is.sort = F)
  de.res = Get_DEG(de.res, "p_value", 0.01, is.sort = F)
  print(head(de.res))
  print(table(substr(rownames(de.res), 1, 3)))
}
system.time(simpleDE.test())

       log_FC      p_value       fdr
DEG1 1.034405 0.0017929311 0.2827846
DEG2 1.340485 0.0008548924 0.2092738
DEG3 1.019590 0.0006365304 0.1917286
DEG4 1.155964 0.0005513130 0.1901079
DEG5 1.071461 0.0037171779 0.4024025
DEG6 1.036673 0.0002875088 0.1505749

DEG non 
 87  32 


   user  system elapsed 
  1.271   0.049   1.320 

In [23]:
# Now we also perform DESeq2 on the same data.
DESeq2.test = function(){
  dds = DESeqDataSetFromMatrix(countData = dt, colData = s_data, design = ~ treatment)
  dds = DESeq(dds)
  res = results(dds)
  res = res[ res$log2FoldChange > log(exp(0.5)) | res$log2FoldChange < -log(exp(0.5)), ]
  res = res[res$pvalue < 0.01, ]
  print(head(res))
  print(table(substr(rownames(res), 1, 3)))
}
# DESeq2.test()
system.time(DESeq2.test())

converting counts to integer mode

  the design formula contains one or more numeric variables with integer values,
  specifying a model with increasing fold change for higher values.
  did you mean for this to be a factor? if so, first convert
  this variable to a factor using the factor() function

estimating size factors

estimating dispersions

gene-wise dispersion estimates

mean-dispersion relationship

final dispersion estimates

fitting model and testing



log2 fold change (MLE): treatment 
Wald test p-value: treatment 
DataFrame with 6 rows and 6 columns
      baseMean log2FoldChange     lfcSE      stat      pvalue        padj
     <numeric>      <numeric> <numeric> <numeric>   <numeric>   <numeric>
DEG1   49.8782        1.48197  0.233326   6.35151 2.13216e-10 3.52787e-08
DEG2   31.5257        1.92298  0.291506   6.59672 4.20340e-11 8.87352e-09
DEG3   50.0395        1.46068  0.255207   5.72350 1.04349e-08 9.82804e-07
DEG4   39.1313        1.65624  0.277419   5.97019 2.36982e-09 2.68667e-07
DEG5   42.8503        1.53544  0.265602   5.78099 7.42645e-09 7.57746e-07
DEG6   41.3594        1.48336  0.275409   5.38604 7.20271e-08 6.21056e-06

DEG non 
100  57 


   user  system elapsed 
  3.584   0.054   3.650 

# Conclusion

In terms of sensitivity, while DESeq2 has a full score, simpleDE has a decent 87% true positive rate.
In terms of specificity, DESeq2 contains a level of (9900-54) / 9900 = 99.45% and simpleDE has a level of (9900 - 33) / 9900 = 99.67%.
Overall, simpleDE is a less conservative method comparing with DESeq2.

In terms of time effiency, simpleDE is much faster than DESeq2 due to simplicity of algorithm and less handling of data structure.